internal/api/http/server/router.go

```diff
@@ -19,82 +20,586 @@ func (s *Server) Run(addr string) error {
+			return
+		}
+
+		// Call GetTaskFiles to retrieve the task files as a .tar.gz archive
+		tarFilePath, err := ts.GetTaskFiles(taskID)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Failed to retrieve task files: %v", err), http.StatusInternalServerError)
+			return
+		}
+		defer utils.RemoveDirectory(tarFilePath)
+
+		// Open the .tar.gz file
+		tarFile, err := os.Open(tarFilePath)
+		if err != nil {
+			http.Error(w, "Failed to open task files archive.", http.StatusInternalServerError)
+			return
+		}
+		defer utils.CloseIO(tarFile)
+
+		// Set headers and serve the .tar.gz file
+		w.Header().Set("Content-Type", "application/gzip")
+		w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=task%dFiles.tar.gz", taskID))
+		w.Header().Set("Content-Length", fmt.Sprintf("%d", utils.FileSize(tarFile)))
+
+		// Stream the file content to the response
+		_, err = io.Copy(w, tarFile)
+		if err != nil {
+			http.Error(w, "Failed to send task files archive.", http.StatusInternalServerError)
+			return
 		}
-		w.Write([]byte("File uploaded successfully"))
 	})
 
-	mux.HandleFunc("/download/", func(w http.ResponseWriter, r *http.Request) {
+	mux.HandleFunc("/getUserSubmission", func(w http.ResponseWriter, r *http.Request) {
 		if r.Method != http.MethodGet {
 			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
 			return
 		}
 
-		fileName := r.URL.Path[len("/download/"):]
-		file, err := os.Open(fmt.Sprintf("./downloaded/%s", fileName))
+		// Extract 'taskID' from query parameters
+		taskIDStr := r.URL.Query().Get("taskID")
+		if taskIDStr == "" {
+			http.Error(w, "taskID is required.", http.StatusBadRequest)
+			return
+		}
+
+		// Extract 'userID' from query parameters
+		userIDStr := r.URL.Query().Get("userID")
+		if userIDStr == "" {
+			http.Error(w, "userID is required.", http.StatusBadRequest)
+			return
+		}
+
+		// Extract 'submissionNumber' from query parameters
+		submissionNumberStr := r.URL.Query().Get("submissionNumber")
+		if submissionNumberStr == "" {
+			http.Error(w, "submissionNumber is required.", http.StatusBadRequest)
+			return
+		}
+
+		// Convert parameters to integers
+		taskID, err := strconv.Atoi(taskIDStr)
+		if err != nil {
+			http.Error(w, "Invalid taskID.", http.StatusBadRequest)
+			return
+		}
+
+		userID, err := strconv.Atoi(userIDStr)
 		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
+			http.Error(w, "Invalid userID.", http.StatusBadRequest)
 			return
 		}
-		defer file.Close()
 
-		// Set the header for the file
-		fileStat, err := file.Stat()
+		submissionNumber, err := strconv.Atoi(submissionNumberStr)
 		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
+			http.Error(w, "Invalid submission number.", http.StatusBadRequest)
 			return
 		}
+
+		// Retrieve the user's submission file content
+		fileContent, fileName, err := ts.GetUserSubmission(taskID, userID, submissionNumber)
```

Comment:
There should be explicit error when user submission doesn't exists. For exaple ErrNotFound, which should be handled properly. Such error should not return InternalServerError Status, but rather 404 Not Found